File: 2026/106628/106628m.cpp
/*!
 * author: manhhomienbienthuy
 * created: 2026-07-30T03:10:37+00:00
 * https://codeforces.com/gym/106628/problem/M
 */

#include <bits/stdc++.h>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int t;
  cin >> t;

  while (t--) {
    int C, Q;
    cin >> C >> Q;

    list<int> pool;

    unordered_map<int, list<int>::iterator> pos;
    unordered_set<int> dirty;

    auto touch = [&](int x) {
      auto it = pos.find(x);

      if (it != pos.end()) {
        pool.erase(it->second);
      }

      pool.push_back(x);
      pos[x] = prev(pool.end());

      if ((int)pool.size() > C) {
        int y = pool.front();
        pool.pop_front();

        pos.erase(y);
        dirty.erase(y);
      }
    };

    while (Q--) {
      int o;
      cin >> o;

      if (o == 1) {
        int x;
        cin >> x;

        touch(x);
      } else if (o == 2) {
        int x;
        cin >> x;

        dirty.insert(x);
        touch(x);
      } else if (o == 3) {
        cout << dirty.size() << '\n';
        dirty.clear();
      } else {
        int x;
        cin >> x;

        cout << (pos.count(x) > 0) << ' ' << (dirty.count(x) > 0) << '\n';
      }
    }
  }

  return 0;
}

File: 2026/106628/106628m.py
# author: manhhomienbienthuy
# created: 2026-07-30T05:22:36+00:00
# https://codeforces.com/gym/106628/problem/M

import sys
input = sys.stdin.readline

t = int(input())
for _ in range(t):
    C, Q = map(int, input().split())

    pool = []
    dirty = []

    for i in range(Q):
        data = input().split()
        if len(data) == 1:
            o, x = int(data[0]), 0
        else:
            o, x = map(int, data)

        if o == 1:
            if x in pool:
                pool.remove(x)
            pool.append(x)
            if len(pool) > C:
                if pool[0] in dirty:
                    dirty.remove(pool[0])
                del pool[0]
            continue
        if o == 2:
            if x not in dirty:
                dirty.append(x)
            if x in pool:
                pool.remove(x)
            pool.append(x)
            if len(pool) > C:
                if pool[0] in dirty:
                    dirty.remove(pool[0])
                del pool[0]
            continue
        if o == 3:
            print(len(dirty))
            dirty.clear()
            continue
        if o == 4:
            print(1 if x in pool else 0, 1 if x in dirty else 0)
            continue
